src/Compilers/Core/Portable/CodeGen/TokenMap.cs

```diff
@@ -24,74 +23,91 @@ namespace Microsoft.CodeAnalysis.CodeGen
         {
-            Debug.Assert(item is Cci.ISignature || item is Cci.IReference);
             uint token;
-
             // NOTE: cannot use GetOrAdd here since items and itemToToken must be in sync
             // so if we do need to add we have to take a lock and modify both collections.
-            lock (_items)
+            lock (_itemEquivalentToToken)
             {
-                if (!_itemToToken.TryGetValue(item, out token))
+                // Check if there is an equivalent type that has a token
+                if (!_itemEquivalentToToken.TryGetValue(item, out token))
                 {
-                    token = (uint)_items.Count;
-                    _items.Add(item);
-                    _itemToToken.Add(item, token);
+                    token = (uint)_count;
+                    // No equivalent, add the token for this type
+                    _itemEquivalentToToken.Add(item, token);
+
+                    var count = (int)token + 1;
+                    var items = _items;
+                    if (items.Length > count)
+                    {
+                        items[(int)token] = item.AsObject();
+                    }
+                    else
+                    {
+                        // Not enough room, we need to resize the array
+                        Array.Resize(ref items, Math.Max(8, count * 2));
+                        items[(int)token] = item.AsObject();
+
+                        // Update the updated array reference before updating _count
+                        Volatile.Write(ref _items, items);
+                    }
+
+                    Volatile.Write(ref _count, count);
                 }
             }
 
+            // Use the provided token to update the reference dictionary
             referenceAdded = _itemIdentityToToken.TryAdd(item, token);
             return token;
         }
 
         public object GetItem(uint token)
         {
-            lock (_items)
-            {
-                return _items[(int)token];
-            }
-        }
-
-        public IEnumerable<object> GetAllItems()
-        {
-            lock (_items)
-            {
-                return _items.ToArray();
-            }
+            return _items[(int)token];
         }
 
         //TODO: why is this is called twice during emit?
         //      should probably return ROA instead of IE and cache that in Module. (and no need to return count)
-        public IEnumerable<object> GetAllItemsAndCount(out int count)
+        public object[] GetAllItems()
         {
-            lock (_items)
-            {
-                count = _items.Count;
-                return _items.ToArray();
-            }
+            // Read the count prior to getting the array
+            int count = Volatile.Read(ref _count);
+            object[] items = Volatile.Read(ref _items);
+
+            // Return a right sized copy of the array
+            return (new ReadOnlySpan<object>(items, 0, count)).ToArray();
```

Comment:
Rather than copying the array, couldn't we just return the ROS? It looks like there's only one caller of this method, and only one caller of the method that calls the method that calls this.

Reply:
Seems fair